plugins/inputs/sqlserver/sqlserver.go

```diff
@@ -244,11 +247,36 @@ func (s *SQLServer) Gather(acc telegraf.Accumulator) error {
 }
 
 func (s *SQLServer) gatherServer(server string, query Query, acc telegraf.Accumulator) error {
-	// deferred opening
-	conn, err := sql.Open("mssql", server)
-	if err != nil {
-		return err
+	// deferred connection opening
+	var conn *sql.DB
+
+	if strings.Contains(strings.ToLower(server), strings.ToLower("password=")) {
+		// when password is provided in connection string, use SQL auth
+
+		var err error
+		conn, err = sql.Open("mssql", server)
+
+		if err != nil {
+			return err
+		}
+	} else {
+		// when password is not provided in connection string, use AAD auth with an MSI token
+
+		tokenProvider, err := getMSITokenProvider()
```

Comment:
I believe for all servers under a single VM, token will be same and we can re-use. Will fetching of token happen once per VM or is this once per server?  If it is latter, is it possible to fetch the token once and re-use for all servers in one VM?

Reply:
Added in-memory token caching mechanism, it gets a token for the managed identity and use it until it finds it expired.